### PR TITLE
Inherit CSS-classes from parent entity

### LINKED ISSUE #158: Inherit CSS-classes from parent entity
If we have `A` with `A_m1_v1`, then inherit `B` from `A` and  set `B_m1_v1` we would like to get CSS-classes: `A A_m1_v1 B B_m1_v1`. Now we have `B B_m1_v1` only.

### Proposed solution

We can wrap every `mods` field in `decl` time to embed info about entity into each modifier. Now we have a hash of modifiers without entities, but we wanna have for parent `A`:
**before**
``` js
mix: { block : 'A', mods : { m1 : 'v1' } } 
```

**after wrap result will be:**
``` js
{ m1 : { val : 'v1', entites : ['A'] } }
```

For child `B`:
**before**
``` js
mods() {
  return { m2 : 'v2' };
}

mix() {
  return { block : 'A', mods : { m1 : 'v1' } };
}
```

**after**
``` js
mods() {
  return {
    ...this.__base(), // { m1 : { val : 'v1', entites : ['A'] } }
    m2 : 'v2'
}
// not need self `mix`

// after wrap result will be
{ m1 : { val : 'v1', entites : ['A', 'B'] }, m2 : { val : 'v2', entities : ['B'] }
```

So we infer all needed CSS-classes from such `mods` structure.

### PR DESCRIPTION
**Pull request by awinogradov**
Close #158

### CODE REVIEW

**Comment by awinogradov** [comment]
- [ ] - rename `cssCollector`
- [ ] - ability to cancel this behavior from `decl`?

**Comment by awinogradov** [comment]
Relates to https://github.com/bem/bem-react-core/issues/182

**Comment by awinogradov** [comment]
Review, plz:
- https://github.com/bem/bem-react-core/pull/180/files?diff=unified#diff-cfbeced5c38e4b6c3c16c872c3438f8aR36 до 61
- tests/bemClassName.spec.js
- https://github.com/bem/bem-react-core/pull/180/files?diff=unified#diff-aa4430dc95d9d108131cee6c66086140R39 до 51
- https://github.com/bem/bem-react-core/pull/180/files?diff=unified#diff-aa4430dc95d9d108131cee6c66086140R242 до 269

File: src/Bem.js [hunk 1]
```diff
@@ -14,95 +15,61 @@ const bemModes = {
...
 
             return ({ addBemClassName = true, block, mods, elem, elemMods, mix, cls }) => {
                 if(addBemClassName) {
-                    const entities = [{ block, elem }],
-                        realMods = getMods({ block, mods, elem, elemMods });
-                    realMods && Object.keys(realMods).forEach(name => {
-                        const val = realMods[name];
-                        val && entities.push({ block, elem, mod : { name, val } });
-                    });
+                    const realMods = resolveMods({ block, mods, elem, elemMods }),
```
Comment (veged) [inline src/Bem.js]: realMods -> resolvedMods?

File: src/Bem.js [hunk 2]
```diff
@@ -146,13 +113,66 @@ export default function({ preset, naming }) {
                     cls && entities.push(cls);
 
                     return entities.map(entity => typeof entity === 'string'?
-                        entity : str(entity)).join(' ');
+                        entity : stringifyEntity(entity)).join(' ');
```
Comment (veged) [inline src/Bem.js]: stringifyEntity -> Entity.stringify ?

File: src/Core.js [hunk 3]
```diff
@@ -1,226 +1,220 @@
...
+            wrapWithFunction(obj,
+                ['addBemClassName', 'tag', 'attrs', 'style', 'content', 'cls', 'mods', 'mix', 'addMix']),
+        collectMods = (fields) => {
+            if(fields.hasOwnProperty('mods')) {
+                const val = fields['mods'];
+                fields['mods'] = function() {
+                    // FIXME: @dfilatov
+                    const collected = [val, 'this.__base'][0].apply(this, arguments) || {};
+                    const entities = collected.__entities || (collected.__entities = {});
+
```
Comment (veged) [inline src/Core.js]: double `const`

File: src/Entity.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const delim = '$';
+
```
Comment (veged) [inline src/Entity.js]: why ‘$’, not ‘-‘?

**Comment by veged** [comment]
only tiny details to fix — and merge!

File: src/Bem.js (hunk 2 above)
Comment (awinogradov) [reply src/Bem.js]: No, its stringify for CSS class.  `Entity.stringify` is a key.

File: src/Entity.js (hunk 4 above)
Comment (awinogradov) [reply src/Entity.js]: `Entity.stringify` is a key. It doesn't know about naming.

File: src/Bem.js (hunk 2 above)
Comment (awinogradov) [reply src/Bem.js]: I'm going to rename it to `tokenize`

**Pull request merged**